source/MaterialXRender/Handlers/ImageHandler.h

```diff
@@ -3,60 +3,173 @@
+class ImageDesc
+{
+  public:
+    /// Image width
+    unsigned width = 0; 
+    /// Image height
+    unsigned height = 0;
+    /// Number of channels
+    unsigned int channelCount = 0;
+    /// Number of mip map levels
+    unsigned int mipCount = 0;
+    /// CPU buffer. May be empty
+    float* resourceBuffer = nullptr;
+    /// Hardware target dependent resource identifier. May be empty
+    unsigned int resourceId = 0;
+
+    /// Compute the number of mip map levels based on size of the image
+    void computeMipCount()
+    {
+        mipCount = (unsigned int)std::log2(std::max(width, height)) + 1;
+    }
+};
+
+/// Image description cache
+using ImageDescCache = std::unordered_map<std::string, ImageDesc>;
+
+/// Shared pointer to an ImageLoader
+using ImageLoaderPtr = std::shared_ptr<class ImageLoader>;
+
+/// @class @ImageLoader
+/// Abstract class representing an disk image loader
+///
+class ImageLoader
+{
+  public:
+    /// Default constructor
+    ImageLoader() {}
+
+    /// Default destructor
+    virtual ~ImageLoader() {}
+
+    /// Returns if a given file extension is supported. 
+    /// @param extension File extenstion to check.
+    virtual bool supportsExtension(const std::string extension) = 0;
+
+    /// Save image to disk. This method must be implemented by derived classes.
+    /// @param fileName Name of file to save image to
+    /// @param imageDesc Description of image
+    /// @return if save succeeded
+    virtual bool saveImage(const std::string& fileName,
+                           const ImageDesc &imageDesc) = 0;
+
+    /// Load an image from disk. This method must be implemented by derived classes.
+    /// @param fileName Name of file to load image from
+    /// @param imageDesc Description of image updated during load.
+    /// @return if load succeeded
+    virtual bool loadImage(const std::string& fileName,
+                           ImageDesc &imageDesc) = 0;
+};
+
 /// Shared pointer to an ImageHandler
 using ImageHandlerPtr = std::shared_ptr<class ImageHandler>;
 
 /// @class @ImageHandler
-/// Abstract class representing an disk image handler
+/// A image handler class. Keeps track of images which are loaded
+/// from disk via supplied ImageLoader. Derive classes are responsible
+/// for determinine how to perform the ligic for "binding" of these resources
+/// for a given target (such as a given shading language).
 ///
 class ImageHandler
 {
-public:
-    /// Default constructor
-    ImageHandler() {}
+  public:
+    /// Constructor. Assume at least one loader must be supplied.
+    ImageHandler(ImageLoaderPtr imageLoader)
+    {
+        _imageLoaders.push_back(imageLoader);
+    }
+
+    /// Add additional image loaders. Useful to handle different file
+    /// extension
+    /// @param loader Loader to add to list of available loaders.
+    void addLoader(ImageLoaderPtr loader)
+    {
+        _imageLoaders.push_back(loader);
```

Comment:
An alternative would be to store the image loaders in a map, using the extension as key. It seems to use linear search and pick the first one in the array with a matching extension. So using a map would be more efficient.

Note that for image formats with multiple extensions (like "jpg" and "jpeg") we could register the same loader twice for both extension variations.

Reply:
I did think about this but thought the map means you must be explicit during registration time which codecs to support where as in a plug-in loader system this may not be known. Formats within codecs may not be known as well. e.g. DDS is an extension name but had various formats.